plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/ForceApply.hs

```diff
@@ -0,0 +1,186 @@
+{- | The 'ForceApply' optimisation pushes 'Force' inside its direct 'Apply' subterms,
+ removing any 'Delay' at the top of the body of the underlying lambda abstraction.
+ For example, @force ((\x -> delay b) a)@ is transformed into @(\x -> b) a@.
+ In such simple cases, the transformation is obviously correct, the question remains
+ if this approach can be generalised.
+
+ Since UPLC programs are created from erasing the types of TPLC programs (see
+ "PlutusCore.Compiler.Erase") we will consider TPLC terms of the following structure:
+
+ @
+   /\T1 -> \X1 -> /\T2 -> \X2 -> /\T3 -> \X3 -> ... -> /\Tn -> \Xn -> body
+ @
+ where @T1 ... Tn@ are lists of type variables (e.g. @T1@ could be @[t, q, p]@)
+ and @X1 ... Xn@ are lists of term variables. Of course, each @/\@ and @\@ here would
+ desugar to a sequence of type/term abstractions.
+
+ In order to reason about the proposed optimisation we need to consider such terms in
+ the context of them being applied to some sequence of terms.
+
+ One important observation is that this transformation requires that the underlying
+ (term)-lambda abstraction will be fully reduced by the applications.
+ For UPLC, this can happen only when the number of lambda abstracted variables is equal
+ to the number of terms to which it will be applied.
+ For example, @force (\x -> delay b) => (\x -> b)@ is invalid, since the former is @error@.
+ The other case, we can see that applying the optimisation modifies the end result:
+ @
+   force ((\x -> delay b) a1 a2) => (\x -> b) a1 a2 => b[x1/a1] a2
+
+   vs.
+
+   force ((\x -> delay b) a1 a2) => force ((delay b[x1/a1]) a2)
+ @
+ The latter cannot be reduced further according to the reducation semantics of UPLC.
+
+ To generalise, we consider the family of terms above applied to a family of types and
+ terms:
+ @
+   (/\T1 -> \X1 ->  ... -> /\Tn -> \Xn -> body)
+     T1 X1 ... Tn Xn
```

Comment:
Overloading of the names is kinda unfortunate. E.g. `T1` means both an expression and a variable mapping to that expression.

Reply:
> Hmm, do you think it would be better to switch to using unicode symbols? Haddock supports unicode, right?

I'd say let's not do unicode in non-agda files. You can simply use lowercase `t` for type variables and uppercase `T` for types for example. Anyway, I'm just nitpicking, feel free to ignore.